File: gene_finding.py
#This program will take a DNA string as input, and return the first ORF within the string

#PSEUDOCODE

#GENE_FINDING(s)
#    stop_codons <- ["TGA", "TAG", "TAA"]
#    s <- s[s.find('ATG'):]
#    for stop_codon in stop_codons do
#        for i in range|s, 3| do
#            if s[i:i+3] = stop_codon then
#                orf <- s[:i+3]
#    return orf

import sys

stop_codons = ["TGA", "TAG", "TAA"]
def gene_finding(s):
    s = s[s.find('ATG'):]
    for stop in stop_codons:
        for j in range(0, len(s), 3):
            if s[j:j+3] == stop:
                s = s[:j+3]
    return s

print(gene_finding(sys.stdin.readline().strip()))

File: Week 2/X66236_en/gene_finding.cpp
/*
This program will take a DNA string as input, and return the first ORF within the string

PSEUDOCODE

GENE_FINDING(s)
    Let stop codons be a list of strings
    Let orf be and empty string
    stop_codons <- ["TGA", "TAG", "TAA"]
    s <- s[s.find('ATG'):]
    for stop_codon in stop_codons do
        for i = 0 to i = |s| every 3 numbers do
            if s[i:i+3] = stop_codon then
                orf <- s[:i+3]
    return orf
*/

#include <iostream>
#include <string>
using namespace std;

int main() {
	string inp;
	cin >> inp;
	for (int i = 0; i < (inp.length()-3) ; i++){
		if (inp.substr(i,3) == "ATG"){
			string ns = inp.substr(1,inp.length());
			for (int j = +3; j < inp.length()-3 ; j+=3){
				if (inp.substr(j,3) == "TAG" or inp.substr(j,3) == "TAA" or inp.substr(j,3) == "TGA"){
					int end = j+3;
					cout << inp.substr(i,end) << endl;
					break;
				}
			}
			break;
		}
	}	
return(0);
}
